File: hphp/hack/src/hhbc/emit_memoize_helpers.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Core_kernel
open Instruction_sequence

let memoize_suffix = "$memoize_impl"

let getmemokeyl local index name =
  [
    instr_getmemokeyl (Local.Named name);
    instr_setl (Local.Unnamed (local + index));
    instr_popc;
  ]

let param_code_sets params local =
  gather
  @@ List.concat_mapi params (fun index param ->
         getmemokeyl local index @@ Hhas_param.name param)

let param_code_gets params =
  gather
  @@ List.map params (fun param ->
         instr_cgetl (Local.Named (Hhas_param.name param)))

let check_memoize_possible pos ~params ~is_method =
  if (not is_method) && List.exists params (fun p -> p.Aast.param_is_variadic)
  then
    Emit_fatal.raise_fatal_runtime
      pos
      "<<__Memoize>> cannot be used on functions with variable arguments"

File: hphp/hack/src/hhbc/emit_memoize_helpers.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use emit_fatal_rust as emit_fatal;
use hhas_param_rust::HhasParam;
use instruction_sequence::{instr, InstrSeq, Result};
use oxidized::{aast::FunParam, pos::Pos};

pub const MEMOIZE_SUFFIX: &str = "$memoize_impl";

pub fn get_memo_key_list(local: local::Id, index: usize, name: String) -> Vec<InstrSeq> {
    vec![
        instr::getmemokeyl(local::Type::Named(name)),
        instr::setl(local::Type::Unnamed(local + index)),
        instr::popc(),
    ]
}

pub fn param_code_sets(params: &[HhasParam], local: local::Id) -> InstrSeq {
    InstrSeq::gather(
        params
            .iter()
            .enumerate()
            .map(|(i, param)| get_memo_key_list(local, i, param.name.clone()))
            .flatten()
            .collect(),
    )
}

pub fn param_code_gets(params: &[HhasParam]) -> InstrSeq {
    InstrSeq::gather(
        params
            .iter()
            .map(|param| instr::cgetl(local::Type::Named(param.name.clone())))
            .collect(),
    )
}

pub fn check_memoize_possible<Ex, Fb, En, Hi>(
    pos: &Pos,
    params: &[FunParam<Ex, Fb, En, Hi>],
    is_method: bool,
) -> Result<()> {
    if !is_method && params.iter().any(|param| param.is_variadic) {
        return Err(emit_fatal::raise_fatal_runtime(
            pos,
            String::from("<<__Memoize>> cannot be used on functions with variable arguments"),
        ));
    }
    Ok(())
}
